design/src/rfcs/rfc0023_serialization_and_deserialization.md

```diff
@@ -0,0 +1,90 @@
+<!-- Give your RFC a descriptive name saying what it would accomplish or what feature it defines -->
+RFC: Serialization and Deserialization
+=============
+
+<!-- RFCs start with the "RFC" status and are then either "Implemented" or "Rejected".  -->
+> Status: RFC
+>
+> Applies to: builder structs and other data types that are connected to builder structs
+
+# Overview
+There has been some github issues requesting for the implementation of Serialize/Deserialize traits provided serde crate.  
+At the current state of sdk, user must implement their own solution in order to serialize/deserialize sdk's datatypes.  
+This is a RFC for implementing serde traits.
+
+# Potential Use Cases
+- Testing
+- Reusable conifiguration, infrastructure as a service
+- Logging
```

Comment:
- I think I understand the testing use case.
- For reusable configuration and infrastructure as a service, I'd like to see more specifics on how this change would accomplish that.
- For logging, I'd like to see a comparison with our current logging practices.
  - (Random thought that just occurred to me) We redact sensitive information in logs. We don't want to encourage people to use this feature for logging unless we can still do that or users understand how this affects their security posture.

Reply:
> For reusable configuration and infrastructure as a service, I'd like to see more specifics on how this change would accomplish that.
> For logging, I'd like to see a comparison with our current logging practices. 

Well, I've been creating a new struct that implements serde traits and converts them into builder types.
For example;
```rust
#[derive(Default, serde::Serialize, Deserialize)]
struct Config {
    job_definition: String,
    job_queue: String,
    job_timeout: i32,
}

impl Config {
  fn job_timeout(&self) -> JobTimeout {
    JobTimeout::builder()
        .attempt_duration_seconds(self.job_timeout)
        .build()
  }
  fn name(&self) -> String {
    format!("{}-{}-{}", create_uuid(), date(), name_of_host_computer());
  }
  fn to_batch(&self) {
    let shared_config = aws_config::from_env().load().await;

    let batch_client = aws_sdk_batch::Client::new(&shared_config);
    let res = batch_client
        .submit_job()
        .job_name(self.name().as_str())
        .job_definition(self.job_definition.as_str())
        .job_queue(self.job_queue.as_str())
        .timeout(self.job_timeout())
        .send()
        .await
        .unwrap();

  }
}
```

This is an actual piece of code that comes from one of my tools.

I found this pretty useful since I can just write stuff in rust and sometimes dynamically generate some parameters.  
The downside is that I need to create a struct that converts itself into SDK's builder types.

For logging, I was thinking of a situation where you want to reproduce something;   
As far as I'm concerned, the current logging services that AWS offers are not exactly designed for that purpose.   

I didn't put any thoughts about the sensitive info;  
I do think that there are security concerns, but I think this is something that programmers should manage.  
Even without `serde` all those data can be extracted and stored somewhere.

Maybe we might want to mention potential danger in the document.

Reply:
At AWS, we have a concept we call "working backwards." We think about our customer's needs, and then come up with solutions to meet that need. This RFC currently reads a bit more like a solution in search of a problem. I think that this RFC would be improved by starting with the customer issues that it solves, including some example code of how customers will use your new feature to solve the problem.

There are multiple solutions to most problems, and by focusing on one, we do a disservice to customers. In your specific case, we already have several customer issues that you linked in the description of this PR. I think it would be good to demonstrate, in the RFC itself, how your proposal would meet the needs of those customers.

Also, while you did mention the tradeoffs of your proposal, that information would be even better contextualized if you could directly link it to those customer needs. Customers don't want "a cure that's worse than the disease", but they may be willing to increase compile times in CI if it makes writing and understanding tests more convenient.

> I do think that there are security concerns, but I think this is something that programmers should manage.

We also want things to be "secure by default." If doing something has security implications, we'll need to do our best to make that clear before customers start doing it. It's less about making the SDK completely safe to use and more about making it easy for customers to understand the risk of taking an action.

I know that's a lot to take in, I hope it helps! Let me know if you want clarification on anything.